extprocconfig/extprocconfig.go

```diff
@@ -0,0 +1,121 @@
+//	  headers:
+//	  - name: x-model-name
+//	    value: gpt4.4444
+//
+// where the input of the external processor is in the OpenAI schema, the model name is populated in the header x-model-name,
+// The model name header `x-model-name` is used in the header matching to make the routing decision. **After** the routing decision is made,
+// the selected backend name is populated in the header `x-backend-name`. For example, when the model name is `llama3.3333`,
+// the request is routed to either backends `kserve` or `awsbedrock` with weights 1 and 10 respectively, and the selected
+// backend, say `awsbedrock`, is populated in the header `x-backend-name`.
+//
+// From Envoy configuration perspective, configuring the header matching based on `x-backend-name` is enough to route the request to the selected backend.
+// That is because the matching decision is made by the external processor and the selected backend is populated in the header `x-backend-name`.
+type Config struct {
+	// InputSchema specifies the API schema of the input format of requests to the external processor.
+	InputSchema VersionedAPISchema `yaml:"inputSchema"`
+	// ModelNameHeaderKey is the header key to be populated with the model name by the external processor.
+	ModelNameHeaderKey string `yaml:"modelNameHeaderKey"`
+	// BackendRoutingHeaderKey is the header key to be populated with the backend name by the external processor
+	// **after** the routing decision is made by the external processor using Rules.
+	BackendRoutingHeaderKey string `yaml:"backendRoutingHeaderKey"`
+	// Rules is the routing rules to be used by the external processor to make the routing decision.
+	// Inside the routing rules, the header ModelNameHeaderKey may be used to make the routing decision.
+	Rules []RouteRule `yaml:"rules"`
+}
+
+// VersionedAPISchema corresponds to LLMAPISchema in api/v1alpha1/api.go.
+type VersionedAPISchema struct {
+	// Schema is the API schema.
+	Schema APISchema `yaml:"schema"`
+	// Version is the version of the API schema. Optional.
+	Version string `yaml:"version,omitempty"`
+}
+
+// APISchema corresponds to APISchema in api/v1alpha1/api.go.
+type APISchema string
+
+const (
+	APISchemaOpenAI     APISchema = "OpenAI"
+	APISchemaAWSBedrock APISchema = "AWSBedrock"
+)
+
+// HeaderMatch is an alias for HTTPHeaderMatch of the Gateway API.
+type HeaderMatch = gwapiv1.HTTPHeaderMatch
+
+// RouteRule corresponds to LLMRouteRule in api/v1alpha1/api.go
+// besides the `Backends` field is modified to abstract the concept of a backend
+// at Envoy Gateway level to a simple name.
+type RouteRule struct {
+	// Headers is the list of headers to match for the routing decision.
+	// Currently, only exact match is supported.
+	Headers []HeaderMatch `yaml:"headers"`
+	// Backends is the list of backends to which the request should be routed to when the headers match.
+	Backends []Backend `yaml:"backends"`
+}
+
+// Backend corresponds to LLMRouteRuleBackendRef in api/v1alpha1/api.go
```

Comment:
If I understand correctly the goal here for the config is not intended to be tied with AI concepts.

Reply:
well, I think not exactly - I assume this will have a semantics-caching, prompt guard etc happening in extproc